zk-token-sdk/src/curve25519/edwards.rs

```diff
@@ -0,0 +1,265 @@
+
+            let result = &left_point + &right_point;
+            Some((&result).into())
+        }
+
+        fn subtract(left_point: &Self, right_point: &Self) -> Option<Self> {
+            let left_point: EdwardsPoint = left_point.try_into().ok()?;
+            let right_point: EdwardsPoint = right_point.try_into().ok()?;
+
+            let result = &left_point - &right_point;
+            Some((&result).into())
+        }
+
+        #[cfg(not(target_arch = "bpf"))]
+        fn multiply(scalar: &PodScalar, point: &Self) -> Option<Self> {
+            let scalar: Scalar = scalar.into();
+            let point: EdwardsPoint = point.try_into().ok()?;
+
+            let result = &scalar * &point;
+            Some((&result).into())
+        }
+    }
+
+    impl MultiScalarMultiplication for PodEdwardsPoint {
+        type Scalar = PodScalar;
+        type Point = Self;
+
+        fn multiscalar_multiply(scalars: Vec<&PodScalar>, points: Vec<&Self>) -> Option<Self> {
+            EdwardsPoint::optional_multiscalar_mul(
+                scalars.into_iter().map(Scalar::from),
+                points
+                    .into_iter()
+                    .map(|point| EdwardsPoint::try_from(point).ok()),
+            )
+            .map(|result| PodEdwardsPoint::from(&result))
+        }
+    }
+}
+
+#[cfg(target_arch = "bpf")]
+mod target_arch {
+    use {
+        super::*,
+        crate::curve25519::curve_syscall_traits::{sol_curve_validate_point, CURVE25519_EDWARDS},
+    };
+
+    pub fn validate_edwards(point: &PodEdwardsPoint) -> Option<bool> {
+        let mut validate_result = 0u8;
+        let result = unsafe {
+            sol_curve_validate_point(
+                CURVE25519_EDWARDS,
+                &point.0 as *const u8,
+                &mut validate_result,
+            )
+        };
+
+        if result == 0 {
+            Some(validate_result == 0)
+        } else {
+            None
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        crate::curve25519::scalar::PodScalar,
+        curve25519_dalek::{
+            constants::ED25519_BASEPOINT_POINT as G, edwards::EdwardsPoint, scalar::Scalar,
+            traits::Identity,
+        },
+        rand::rngs::OsRng,
+    };
+
+    #[test]
+    fn test_validate_edwards() {
+        let pod = PodEdwardsPoint(G.compress().to_bytes());
+        assert!(validate_edwards(&pod).unwrap());
+
+        let invalid_bytes = [
+            120, 140, 152, 233, 41, 227, 203, 27, 87, 115, 25, 251, 219, 5, 84, 148, 117, 38, 84,
+            60, 87, 144, 161, 146, 42, 34, 91, 155, 158, 189, 121, 79,
+        ];
+
+        assert!(!validate_edwards(&PodEdwardsPoint(invalid_bytes)).unwrap());
+    }
+
+    #[test]
+    fn test_edwards_add_subtract() {
+        // identity
+        let identity = PodEdwardsPoint(EdwardsPoint::identity().compress().to_bytes());
+        let random_point = PodEdwardsPoint((Scalar::random(&mut OsRng) * G).compress().to_bytes());
```

Comment:
consider fixed inputs to avoid flaky tests.  if we need to fuzz, we should do that explicitly